src/twisted/_version.py

```diff
@@ -7,5 +7,5 @@
 
 from incremental import Version
 
-__version__ = Version('Twisted', 17, 1, 0, dev=0)
+__version__ = Version('Twisted', 17, 1, 1, dev=0)
```

Comment:
> In what circumstance would you find yourself with a random git revision of Twisted trunk as an installable option?

It's more likely that it would be an _already installed_ version and pip would leave it there as "good enough".  This might happen in the course of trying to integrate Twisted fixes that were developed downstream and need to be pushed upstream (so I want to have a development version of Twisted in my environment to verify the fixes still work).  Once there's a workflow that involves installing a development version of Twisted, I'd really like the tools to protect me from user error.  I don't know about other people but I have a virtualenv working-set size of about twenty and sometimes I get confused about what's supposed to be in different environments (and sometimes I screw up and do something that _puts_ the wrong thing in some of them).

> If you did that, you'd also find that pip doesn't include a constraint solver and will happily violate your previously-specified == and >= constraints.

Yes.  Unfortunately, the safeguards I want don't _really_ exist since the kind of version checks that cause "problems" for entrypoints don't really get applied anywhere else - once a virtualenv has had the wrong version of things installed, oftentimes you're just screwed and it's up to you, the user, to notice and fix things.  However, such checks _could_ eventually be done.  And I could also invoke a manual check myself, for example as part of my test runner startup.

Regarding `--pre`, I don't think it's much relevant given a pre release could be installed explicitly.  However, I'll note that I _do_ often use `--pre` because I have _other_ requirements that I sometimes need a pre release of.  Since `--pre` isn't scoped, once I have to use it for one project I'm at risk for getting a pre release of anything.  So this is another reason to make sure any pre-releases that _might_ get discovered (but I'm not claiming these versions of Twisted will be discoverable in my particular workflow) at least compare correctly against other versions.

> That said: we could just say that the official release version to go on trunk is always just "always increment the minor release number, unless the previous release is 12, in which case increment the major and reset the minor to 1", as this would address your concern about pinning below a particular release, and even if we get the release prediction wrong, it doesn't cause any version specifiers to compare inaccurately.

I think that what's most surprising about push-back to my push-back (I don't know if I would have commented in the first place if I'd know how far this thread would eventually extend) is that the change you suggest here is just as trivial as the change proposed by this PR and it is objectively (it's tempting to say "obviously" as well but perhaps that's just due to my personal perspective) more correct.  It has the small shortcoming that you point out but this seems like a non-issue to me.